lib/IPC.cpp

```diff
@@ -0,0 +1,266 @@
+/** @file IPC.cpp
+ *  @author Bao Vu (bao.vu@unikie.com)
+ *  @brief Include IPC class methods' definition & non-member functions' definition
+ */
+
+#include "lib/IPC.h"
+
+/******************************************************************************/
+/* IPC Class's methods ********************************************************/
+/******************************************************************************/
+
+unsigned long IPC::get_file_size() const
+{
+	return boost::filesystem::file_size(this->opts.file_name);
+}
+
+void IPC::cleanup()
+{
+	if (this->buffer != nullptr)
+	{
+		delete buffer;
+		std::cout << "Freed the buffer." << std::endl;
+	}
+	std::cout << "Closed file." << std::endl;
+	fs.close();
+}
+
+void IPC::print_members() const
+{
+	std::cout << "Mode: " << static_cast<std::underlying_type<IPCMode>::type>(opts.mode) << "\n"
+			  << "Protocol: " << static_cast<std::underlying_type<IPCProtocol>::type>(opts.protocol) << "\n"
+			  << "server_name: " << opts.server_name << "\n"
+			  << "file_name: " << opts.file_name << "\n"
+			  << "mem_size: " << opts.mem_size << "\n"
+			  << "timeout: " << timeout << "\n";
+	if (this->buffer == nullptr)
+		std::cout << "buffer: nullptr" << std::endl;
+	else
+		std::cout << "buffer: " << buffer << std::endl;
+}
+
+void IPC::auto_start()
+{
+	this->init();
+	if (this->opts.mode == IPCMode::SEND_MODE)
+		this->send();
+	if (this->opts.mode == IPCMode::RECEIVE_MODE)
+		this->receive();
+}
+
+void IPC::open_file()
+{
+	if (this->opts.mode == IPCMode::SEND_MODE)
+		this->fs.open(opts.file_name, std::fstream::in | std::fstream::binary);
+	if (this->opts.mode == IPCMode::RECEIVE_MODE)
+		this->fs.open(opts.file_name, std::fstream::out | std::fstream::trunc | std::fstream::binary);
+
+	if (this->fs.fail()) // Check logical error on I/O operation (close/open)
+		throw std::runtime_error("ERROR: " + this->opts.file_name + ": " + strerror(errno));
```

Comment:
When you throw without catching, you can be sure all destructors will be called. I recommend you to catch in the main for instance.

Reply:
spot on. I will add cleanup() before throw instead. I want to keep my main() clean :D

Reply:
I do have destructor.